### DISCUSSION TITLE
"Nested or scoped quantifiers" in Matcher documentation

### DISCUSSION THREAD

**Discussion by slbayer** [explosion/spaCy#8420] [Help: Coding & Implementations] [answered]
In the matcher documentation, regarding the OP element, it says: "Note that there are no nested or scoped quantifiers – instead, you can build those behaviors with on_match callbacks." What does this mean? Can someone provide an example? Thanks.

**Comment by polm** [comment, marked as answer]
A nested quantifier with regexes would be like `(a*b)*`, where a star expression is applied to something containing a star expression (or other quantifier). This might be useful for checking some kinds of duplication or drawn out words (like `haaaaaahahahaha` or something). I'm not exactly sure what a scoped quantifier would be in this context, but it could be the same as a nested quantifier, or maybe something like "match this pattern two to four times". 

You can't express these using the OP settings, but since you can use `on_match` callbacks to do anything, you can use those to simulate those features.

**Reply by slbayer** [reply]
Thanks for responding. I was pretty sure that's what was meant by nested quantifiers, but I'm not sure how you can build that behavior with on_match. Let's say I want to match the pattern `(a*b)*`, so `a*b` zero or more times. Once I have a match for that rule, how do I match it again _starting at that index_? If the Matcher provided the equivalent behavior to re.match() vs. re.search(), it would be trivial; but the only thing I can think of doing is creating a Span over the next N elements of the Doc (where N is the length of the matched rule, so 2 here) and running the matcher again and seeing if I get a match at the beginning of the Span. And that only works in some circumstances, since you wouldn't be able to do this for non-peripheral token sequences. What am I missing?

**Reply by polm** [reply]
I don't think you're missing anything, I think it's just a case we haven't provided much for. The explanation that you can use `on_match` callbacks isn't an assertion that's the best possible way, just a note about a workaround if you need something we haven't implemented yet. Because the matcher operates on tokens we have to implement it from scratch and be careful about performance. 

I would note that because the matcher gives you access to the original document, you can always use it to identify candidate docs/spans and run a normal regex on the result, which can be better than running the matcher recursively depending on what you're trying to match.

Also, when writing regexes *please* use inline Markdown - if you don't you end up with mysterious alternating italics.
